storagenode/gracefulexit/worker.go

```diff
@@ -4,48 +4,261 @@
+			if err != nil {
+				return errs.Wrap(err)
+			}
+			break
+		case *pb.SatelliteMessage_ExitCompleted:
+			worker.log.Info("graceful exit completed.", zap.Stringer("satellite ID", worker.satelliteID))
+
+			err = worker.satelliteDB.CompleteGracefulExit(ctx, worker.satelliteID, time.Now(), satellites.ExitSucceeded, msg.ExitCompleted.GetExitCompleteSignature())
+			if err != nil {
+				return errs.Wrap(err)
+			}
+			break
+		default:
+			// TODO handle err
+			worker.log.Error("unknown graceful exit message.", zap.Stringer("satellite ID", worker.satelliteID))
+		}
+
+	}
 
-	err = worker.satelliteDB.CompleteGracefulExit(ctx, satelliteID, time.Now(), satellites.ExitSucceeded, []byte{})
 	return errs.Wrap(err)
 }
 
+func (worker *Worker) handleFailure(ctx context.Context, transferError pb.TransferFailed_Error, pieceID pb.PieceID, send func(*pb.StorageNodeMessage) error) {
+	failure := &pb.StorageNodeMessage{
+		Message: &pb.StorageNodeMessage_Failed{
+			Failed: &pb.TransferFailed{
+				OriginalPieceId: pieceID,
+				Error:           transferError,
+			},
+		},
+	}
+
+	sendErr := send(failure)
+	if sendErr != nil {
+		worker.log.Error("unable to send failure.", zap.Stringer("satellite ID", worker.satelliteID))
+	}
+}
+
 // Close halts the worker.
 func (worker *Worker) Close() error {
 	// TODO not sure this is needed yet.
 	return nil
 }
+
+// TODO This comes from piecestore.Endpoint. It should probably be an exported method so I don't have to duplicate it here.
+func (worker *Worker) getHashAndLimit(ctx context.Context, pieceReader *pieces.Reader, limit *pb.OrderLimit) (pieceHash pb.PieceHash, orderLimit pb.OrderLimit, err error) {
+
+	if pieceReader.StorageFormatVersion() == 0 {
+		// v0 stores this information in SQL
+		info, err := worker.store.GetV0PieceInfoDB().Get(ctx, limit.SatelliteId, limit.PieceId)
+		if err != nil {
+			worker.log.Error("error getting piece from v0 pieceinfo db", zap.Error(err))
+			return pb.PieceHash{}, pb.OrderLimit{}, err
+		}
+		orderLimit = *info.OrderLimit
+		pieceHash = *info.UplinkPieceHash
+	} else {
+		//v1+ stores this information in the file
+		header, err := pieceReader.GetPieceHeader()
+		if err != nil {
+			worker.log.Error("error getting header from piecereader", zap.Error(err))
+			return pb.PieceHash{}, pb.OrderLimit{}, err
+		}
+		orderLimit = header.OrderLimit
+		pieceHash = pb.PieceHash{
+			PieceId:   limit.PieceId,
+			Hash:      header.GetHash(),
+			PieceSize: pieceReader.Size(),
+			Timestamp: header.GetCreationTime(),
+			Signature: header.GetSignature(),
+		}
+	}
+
+	return
+}
+
+// verifyPieceHash verifies whether the piece hash matches the locally computed hash and the piece hash signature is valid.
+func verifyPieceHash(ctx context.Context, limit *pb.OrderLimit, signee signing.Signee, hash *pb.PieceHash, expectedHash []byte) (err error) {
```

Comment:
Remove piece hash verification logic since it is not necessary (based on our convo earlier)

Reply:
Removed